In [2]:
#Import necessary packages
import pandas as pd
import numpy as np
import string

## 1. Load the data into python


In [5]:
#Load csv file
pd.read_csv('data/School_Learning_Modalities__2021-2022_20240309.csv')

FileNotFoundError: [Errno 2] No such file or directory: 'data/School_Learning_Modalities__2021-2022_20240309.csv'

In [10]:
#Load csv file into a dataframe
df = pd.read_csv('D:\Health Informatics\HHA507 - Data Science\Assignments\Assignment 2 - Data Cleaning\School_Learning_Modalities__2021-2022_20240309.csv')

## 2. Print the count of columns and rows


In [11]:
#Display count of rows and columns
df.shape

(994788, 9)

#### We see here that the data contains 9 columns and 994788 rows

## 3. Provide a print out of the column names


In [12]:
#Display column names
df.columns.values.tolist()

['District NCES ID',
 'District Name',
 'Week',
 'Learning Modality',
 'Operational Schools',
 'Student Count',
 'City',
 'State',
 'ZIP Code']

## 4. Clean the column names


In [13]:
#Clean column names
df.columns = df.columns.str.strip()  # Remove leading and trailing whitespaces
df.columns = df.columns.str.lower()  # Convert to lowercase
df.columns = df.columns.str.replace(' ', '_')  # Replace spaces with underscores
df.columns = df.columns.str.replace('[^A-Za-z0-9]+', '_') #Replace all special characters (that we may not see) with underscore
df.columns.values.tolist()

C:\Users\ttsegyal\AppData\Local\Temp\ipykernel_6772\2812968162.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  df.columns = df.columns.str.replace('[^A-Za-z0-9]+', '_') #Replace all special characters (that we may not see) with underscore


['district_nces_id',
 'district_name',
 'week',
 'learning_modality',
 'operational_schools',
 'student_count',
 'city',
 'state',
 'zip_code']

In [14]:
#Check work
df.head()

district_nces_id     district_name                    week  \
0            100005  Albertville City  12/25/2022 12:00:00 AM   
1            100006   Marshall County  12/25/2022 12:00:00 AM   
2            100007       Hoover City  12/25/2022 12:00:00 AM   
3            100008      Madison City  12/25/2022 12:00:00 AM   
4            100011        Leeds City  12/25/2022 12:00:00 AM   

  learning_modality  operational_schools  student_count          city state  \
0         In Person                    6         5824.0   Albertville    AL   
1         In Person                   15         5764.0  Guntersville    AL   
2         In Person                   18        14061.0        Hoover    AL   
3         In Person                   11        11695.0       Madison    AL   
4         In Person                    4         2076.0         Leeds    AL   

   zip_code  
0     35950  
1     35976  
2     35243  
3     35758  
4     35094

## 5. Clean the strings that might exist within each column


In [15]:
# Clean strings in each column
cols = df.select_dtypes(include='object').columns

for col in cols:
    if col != 'week':  # Exclude 'week' column from operations to maintain datetime format
        df[col] = df[col].str.strip() # Remove leading and trailing whitespaces
        df[col] = df[col].str.lower() # Convert to lowercase
        df[col] = df[col].str.replace(' ', '_') # Replace spaces with underscores
        df[col] = df[col].str.replace('[^A-Za-z0-9]+', '_') # Replace all special characters (that we may not see) with underscore

df.head()


C:\Users\ttsegyal\AppData\Local\Temp\ipykernel_6772\3493790585.py:9: FutureWarning: The default value of regex will change from True to False in a future version.
  df[col] = df[col].str.replace('[^A-Za-z0-9]+', '_')


district_nces_id     district_name                    week  \
0            100005  albertville_city  12/25/2022 12:00:00 AM   
1            100006   marshall_county  12/25/2022 12:00:00 AM   
2            100007       hoover_city  12/25/2022 12:00:00 AM   
3            100008      madison_city  12/25/2022 12:00:00 AM   
4            100011        leeds_city  12/25/2022 12:00:00 AM   

  learning_modality  operational_schools  student_count          city state  \
0         in_person                    6         5824.0   albertville    al   
1         in_person                   15         5764.0  guntersville    al   
2         in_person                   18        14061.0        hoover    al   
3         in_person                   11        11695.0       madison    al   
4         in_person                    4         2076.0         leeds    al   

   zip_code  
0     35950  
1     35976  
2     35243  
3     35758  
4     35094

## 6. Assesses white space or special characters


In [16]:
# Assess whitespace and special characters in the data
contains_special_chars = df[cols].apply(lambda x: x.str.contains('[^A-Za-z0-9]+')) # Check for special characters
contains_whitespace = df[cols].apply(lambda x: x.str.contains(' ')) # Check for whitespace

print(contains_special_chars) 
print(contains_whitespace)

        district_name  week  learning_modality   city  state
0                True  True               True  False  False
1                True  True               True  False  False
2                True  True               True  False  False
3                True  True               True  False  False
4                True  True               True  False  False
...               ...   ...                ...    ...    ...
994783           True  True               True  False  False
994784           True  True               True  False  False
994785           True  True              False  False  False
994786           True  True               True   True  False
994787           True  True               True   True  False

[994788 rows x 5 columns]
        district_name  week  learning_modality   city  state
0               False  True              False  False  False
1               False  True              False  False  False
2               False  True              False  False  Fal

## 7. Converts the column types to the correct types

In [17]:
#Check data types of each column
df.dtypes

district_nces_id         int64
district_name           object
week                    object
learning_modality       object
operational_schools      int64
student_count          float64
city                    object
state                   object
zip_code                 int64
dtype: object

#### Week is in 'object' format, so we must change it to datetime

In [18]:
#Convert 'week' column to datetime
df['week'] = pd.to_datetime(df['week']) # Convert to datetime
df.dtypes # Check work

district_nces_id                int64
district_name                  object
week                   datetime64[ns]
learning_modality              object
operational_schools             int64
student_count                 float64
city                           object
state                          object
zip_code                        int64
dtype: object

## 8. Check for duplicate rows and remove duplicate rows


In [19]:
#Check for duplicate rows
df.duplicated() #check for duplicates

0         False
1         False
2         False
3         False
4         False
          ...  
994783    False
994784    False
994785    False
994786    False
994787    False
Length: 994788, dtype: bool

#### Seems like there are no duplicate rows, but will run the following code for good measure anyways.

In [20]:
#Remove duplicate rows
df.drop_duplicates()

district_nces_id                        district_name       week  \
0                 100005                     albertville_city 2022-12-25   
1                 100006                      marshall_county 2022-12-25   
2                 100007                          hoover_city 2022-12-25   
3                 100008                         madison_city 2022-12-25   
4                 100011                           leeds_city 2022-12-25   
...                  ...                                  ...        ...   
994783           5900187            hannahville_indian_school 2021-08-01   
994784           5900190                 tiospaye_topa_school 2021-08-01   
994785           5900193          seba_dalkai_boarding_school 2021-08-01   
994786           5900197                          noli_school 2021-08-01   
994787           7200030  puerto_rico_department_of_education 2021-08-01   

       learning_modality  operational_schools  student_count          city  \
0              in_person                    6         5824.0   albertville   
1              in_person                   15         5764.0  guntersville   
2              in_person                   18        14061.0        hoover   
3              in_person                   11        11695.0       madison   
4              in_person                    4         2076.0         leeds   
...                  ...                  ...            ...           ...   
994783         in_person                    1          163.0        wilson   
994784         in_person                    1          157.0       laplant   
994785            hybrid                    1           65.0       winslow   
994786         in_person                    1          123.0   san_jacinto   
994787         in_person                  849       292518.0      hato_rey   

       state  zip_code  
0         al     35950  
1         al     35976  
2         al     35243  
3         al     35758  
4         al     35094  
...      ...       ...  
994783    bi     49896  
994784    bi     57652  
994785    bi     86047  
994786    bi     92581  
994787    pr       919  

[994788 rows x 9 columns]

## 9. Assess missingness (count of missing values per column)
 

In [21]:
#Check for missing values
df.isnull().sum()


district_nces_id          0
district_name             0
week                      0
learning_modality         0
operational_schools       0
student_count          1015
city                      0
state                     0
zip_code                  0
dtype: int64

#### The student_count column contains missing values so we will replace those empty cells with 'NaN'

In [22]:
#replace empty cells with nan 
df.replace(to_replace='', value=np.nan, inplace=True) 

#### Now we will check if the changes went through

In [23]:
# Check if missing values were replaced with NaN
missing_values = df['student_count'].isnull()

# Display the rows where there are missing values in the 'student_count' column
missing_df = df.loc[missing_values, ['district_name', 'student_count']]
print(missing_df)


                                           district_name  student_count
325     legacy_traditional_school_north_phoenix_1000568_            NaN
1884                            sussex_montessori_school            NaN
3788               warren_easton_charter_foundation_inc_            NaN
4915                      three_rivers_montessori_school            NaN
4917                           rock_ridge_public_schools            NaN
...                                                  ...            ...
989909                     brilla_caritas_charter_school            NaN
989910                         brilla_pax_charter_school            NaN
990602                       achievement_charter_academy            NaN
990899                 central_point_preparatory_academy            NaN
992915                    promesa_academy_charter_school            NaN

[1015 rows x 2 columns]


## 10. New data field: attempt to create a new column called modality_inperson. This column should contain a binary value of true or false. 

In [24]:
#Create new column called modality_inperson. This column contains a binary value of true or false
df['modality_inperson'] = (df['learning_modality'].apply(lambda x: 'true' if x == 'in_person' else 'false'))

In [25]:
#Check the count of each value in the new column and if changes went through
df['modality_inperson'].value_counts()

true     986119
false      8669
Name: modality_inperson, dtype: int64

In [26]:
#Check work
df.head(10)

district_nces_id     district_name       week learning_modality  \
0            100005  albertville_city 2022-12-25         in_person   
1            100006   marshall_county 2022-12-25         in_person   
2            100007       hoover_city 2022-12-25         in_person   
3            100008      madison_city 2022-12-25         in_person   
4            100011        leeds_city 2022-12-25         in_person   
5            100012         boaz_city 2022-12-25         in_person   
6            100030    alexander_city 2022-12-25         in_person   
7            100060    andalusia_city 2022-12-25         in_person   
8            100100         arab_city 2022-12-25         in_person   
9            100120       athens_city 2022-12-25         in_person   

   operational_schools  student_count            city state  zip_code  \
0                    6         5824.0     albertville    al     35950   
1                   15         5764.0    guntersville    al     35976   
2                   18        14061.0          hoover    al     35243   
3                   11        11695.0         madison    al     35758   
4                    4         2076.0           leeds    al     35094   
5                    5         2440.0            boaz    al     35957   
6                    5         3036.0  alexander_city    al     35010   
7                    3         1839.0       andalusia    al     36421   
8                    4         2696.0            arab    al     35016   
9                    8         4505.0          athens    al     35611   

  modality_inperson  
0              true  
1              true  
2              true  
3              true  
4              true  
5              true  
6              true  
7              true  
8              true  
9              true

In [ ]:
#Save cleaned version
df.to_csv('data/School_Learning_Modalities.csv')